### ISSUE TITLE
SCIP index symbols for struct/enum/const/static/impl nested in functions do not include function name

### ISSUE THREAD

**Issue by mgsloan** [rust-lang/rust-analyzer#18771]
This is demonstrated in tests included in #18758

The reason I did not fix this in that PR is that `hir` representation only appears to know what module these definitions are in, and not whether they are contained in a function.  I am not familiar enough with rust-analyzer internals to provide such information.  The fix would involve a change to the `enclosing_definition` method or the methods it calls.

**Comment by ChayimFriedman2** [comment]
The information whether something is inside function (or anything with body) is whether the module is a block module (this exists in hir-def, but adding it to hir should be easy enough). The problem is that block modules usually have no names. Handling it can be done by special-casing block modules to look at the name of the enclosing definitions.

**Comment by mgsloan** [comment]
@ChayimFriedman2 Thanks for the info, glad this is in the representation. I've attempted to write a method to get the info, but got stuck going from `BlockId` to the containing `FunctionId`.  Maybe I could find the function name in the syntax tree...  But it would be much cleaner to get a `FunctionId`.

```
    pub fn function_parent(self, db: &dyn HirDatabase) -> Option<Function> {
        let containing_block = self.id.containing_block()?;
        let node = containing_block.lookup(db).ast_id.file_syntax(db);
    }
```

Would really appreciate it if on of you rust-analyzer experts would solve this part - feel free to push a commit to this branch adding such a function.  I'm happy to make the moniker.rs changes based upon it.

**Comment by Veykril** [comment]
I believe we don't actually have an association from block to its container, only block to its parent module (which may be a block again).

**Comment by ChayimFriedman2** [comment]
Indeed; but we do preserve an association between a block and its `ast::BlockExpr`, and we can use that (`ancestors_with_macros()`) to find the container.

**Comment by mgsloan** [comment]
Thanks for the further info @ChayimFriedman2 !  I tried for about 15 minutes to sort out how to write the code for this by referencing [this seemingly related code](https://github.com/rust-lang/rust-analyzer/blob/34dc94bb2d05def8b88739a9382b127feb1a35a0/crates/hir/src/semantics/source_to_def.rs#L209) I found by searching for `ancestors_with_macros`.

I would prefer to not block merging #18758 on this concern.   It takes SCIP's symbol information from being wildly out of spec to only having a few known flaws.  If we are blocking merge on this concern, then I would really appreciate it if someone that knows rust-analyzer better than me could update `enclosing_definition` on my branch to return functions when things are nested within them, instead of skipping over the function definitions.

**Comment by ChayimFriedman2** [comment]
I don't think we should block on that either.
